File: day06.kt
package day06

import java.io.File

fun part1(p: String): Int {
    return p.windowed(4).indexOfFirst { s -> s.toSet().size == 4 } + 4
}

fun part2(p: String): Int {
    return p.windowed(14).indexOfFirst { s -> s.toSet().size == 14 } + 14
}


fun main() {
    val data = File("inputs/day06.txt").readText()
    val answer1 = part1(data)
    println("Answer 1: $answer1") // 1760
    val answer2 = part2(data)
    println("Answer 2: $answer2") // 2974
}

File: day06.rs
use std::{fs, collections::HashSet};

fn first_unique_window_end(buffer: &String, window_size: usize) -> usize {
    for idx in (window_size - 1)..buffer.len() {
        let mut set: HashSet<char> = HashSet::new();
        let window = buffer.get((idx - (window_size - 1))..=idx).unwrap();
        set.extend(window.chars());
        if set.len() == window_size {
            return idx + 1;
        }
    }
    return usize::MAX;
}

fn part1(buffer: &String) -> usize {
    first_unique_window_end(buffer, 4)
}

fn part2(buffer: &String) -> usize {
    first_unique_window_end(buffer, 14)
}

fn main() {
    let data: String = fs::read_to_string("inputs/day06.txt").unwrap();
    let answer1 = part1(&data);
    println!("Answer 1: {}", answer1);  // 1760
    let answer2 = part2(&data);
    println!("Answer 2: {}", answer2);  // 2974
}